impl/src/main/java/com/okta/sdk/impl/http/authc/OAuth2RequestAuthenticator.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2020-Present Okta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.okta.sdk.impl.http.authc;
+
+import com.okta.commons.http.Request;
+import com.okta.commons.http.authc.RequestAuthenticationException;
+import com.okta.commons.http.authc.RequestAuthenticator;
+import com.okta.sdk.authc.credentials.ClientCredentials;
+import com.okta.sdk.impl.oauth2.OAuth2AccessToken;
+import com.okta.sdk.impl.oauth2.OAuth2ClientCredentials;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.security.InvalidKeyException;
+
+/**
+ * This implementation used by OAuth2 flow adds Bearer header with access token as the
+ * value in all outgoing requests. This has logic to fetch a new token and store it in
+ * {@link OAuth2ClientCredentials} object if the existing one has expired.
+ *
+ * @since 0.6.0
+ */
+public class OAuth2RequestAuthenticator implements RequestAuthenticator {
+    private static final Logger log = LoggerFactory.getLogger(OAuth2RequestAuthenticator.class);
+
+    private final ClientCredentials<OAuth2AccessToken> clientCredentials;
+
+    public OAuth2RequestAuthenticator(ClientCredentials<OAuth2AccessToken> clientCredentials) {
+        this.clientCredentials = clientCredentials;
+    }
+
+    @Override
+    public void authenticate(Request request) throws RequestAuthenticationException {
+
+        if (clientCredentials != null &&
+            clientCredentials.getCredentials() != null) {
+            OAuth2AccessToken oAuth2AccessToken = clientCredentials.getCredentials();
+
+            if (oAuth2AccessToken != null) {
+                if (oAuth2AccessToken.hasExpired()) {
+                    log.debug("OAuth2 access token expiry detected. Will fetch a new token from Authorization server");
+
+                    synchronized(this) {
+                        try {
+                            OAuth2ClientCredentials oAuth2ClientCredentials = (OAuth2ClientCredentials) clientCredentials;
+                            // clear old token
+                            oAuth2ClientCredentials.reset();
```

Comment:
Check this for thread safety, it looks like this object is used outside of the synchronized block, so the state could be changed on other threads.  You may want to avoid calling reset, and just get a new one, and update the object in a single method call.

It looks like there is also an issue the will cause multiple tokens to be fetched once, for threads waiting to enter the `synchronized` block.  To avoid that you could recheck the state of the object `oAuth2AccessToken` after you enter into the block.   This is similar to the typical double-check singleton pattern: https://www.baeldung.com/java-singleton-double-checked-locking

NOTE: that isn't the only way to solve this, I just offering up one idea.

Reply:
Refactored code around this idea. Please take a look.